### DISCUSSION TITLE
How to store state in the app?

### DISCUSSION THREAD

**Discussion by chbndrhnns** [spec-first/connexion#1859] [Q&A]
On startup, I'd like to store custom state inside the app (like settings) which is also available outside of the context of a request. How would I do that? I noticed that the mechanism from Starlette (via `app.state`) seems not to be available.

**Comment by FelixSchwarz** [comment]
Yes, I have basically the same question.

I think currently it boils down to storing that in a global symbol somewhere but that only works if the settings are immutable. Using connexion's scope could be another solution but that is only available after all middlewares did run (#1750).

**Comment by chrisinmtown** [comment]
Are you asking about static configuration that gets read at startup, or something else?  My Connexion projects all use Flask, and all use the Flask configuration feature, basically slurp in a series of `var = "value"` assignments from a `config.py` file. My code plucks values out of that app config while processing requests by referencing the `conn_app.app.config` object, which supports a dict-like `get` operation.

Are you asking if you can change the contents of the config dict /after/ startup? I would have to try, I don't know if the Flask config object becomes immutable.

**Reply by FelixSchwarz** [reply]
Sounds similar to what I was doing but in my case I need to access the flask config in the function referenced by `APIKEYINFO_FUNC`. That function is triggered by the `SecurityMiddleware` and the flask config is not available by that time (iirc).

I solved that problem by isolating our config changes from the flask config and storing it in a global symbol for all threads. That had some other benefits which I why I wanted to do that anyway but in general I'd like to know if Connexion 3 provides a way to get state/settings outside of handlers.

**Reply by chrisinmtown** [reply]
My code creates a connexion.FlaskApp, loads the config into the Flask app, and /then/ hooks the API spec.  

One of my REST servers uses Connexion security with `x-apikeyInfoFunc`. Do any of those functions (security request handlers) access app config? I could check, but I sure don't see why it would be a problem.

**Reply by FelixSchwarz** [reply]
Not sure if I understood you correctly, but my problem was the following:

```python
os.environ['APIKEYINFO_FUNC'] = 'my_pkg.auth.apikey_auth'
app = build_connexion_app(...)

...

def apikey_auth(provided_key, required_scopes, request):
    api_key = config.get('api_secret')

```

`apikey_auth` is via `connexion.middleware.security.SecurityMiddleware` and that runs before the flask context is available - though I would love to be wrong.

**Reply by chrisinmtown** [reply]
Not at all sure I follow, it appears that you are assigning a Python `pkg.module.function` path to an environment variable? I suspect you are doing something far more sophisticated than I am, and my cookie-cutter examples are useless to you, sorry.

**Reply by FelixSchwarz** [reply]
Don't underestimate your advice :-)

The environment variable is strange indeed but is the only way [described in the docs](https://connexion.readthedocs.io/en/latest/security.html#apikey-authentication) if you don't want to add the module reference in the public openapi yaml. So it is really not that advanced, just a somewhat ugly API.

**Reply by chrisinmtown** [reply]
Thanks for explaining. My open API spec uses multiple x-apikey functions (so a single env variable is not enough), and I don't mind exposing the pkg.module.function names.
